core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/KeyLock.scala

```diff
@@ -0,0 +1,64 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+import java.util.concurrent.TimeUnit
+
+import cats.effect.concurrent.Semaphore
+import cats.effect.implicits._
+import cats.effect.{Blocker, ConcurrentEffect, ContextShift}
+import cats.implicits._
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+
+import scala.concurrent.duration.FiniteDuration
+
+/**
+ * A functional lock which works on a per-key basis.
+ */
+abstract class KeyLock[F[_], K] {
+  def acquire(key: K): F[Unit]
```

Comment:
does `acquire` and `release` still needed as public interface?

Reply:
I guess not, we only use `withKeyLock`. I could remove acquire/release